src/smt/proof_manager.cpp

```diff
@@ -119,6 +119,10 @@ constexpr typename std::vector<T, Alloc>::size_type erase_if(
 std::shared_ptr<ProofNode> PfManager::connectProofToAssertions(
     std::shared_ptr<ProofNode> pfn, SmtSolver& smt, ProofScopeMode scopeMode)
 {
+  // Always clone the proof, since we may be processing it in multiple
+  // ways (e.g. if the user gets a proof and gets an unsat core after the same
+  // check-sat).
+  pfn = pfn->clone();
```

Comment:
I'm a little worried that this is overkill and that it may lead to performance issues (we had those in the past when cloning when there were alternatives). Rather than cloning couldn't you save a reference to the proof node that is reworked *if* you get a new request?

This is the approach I followed in #9585 when different kinds of proofs from the propositional engine may be requested.

Reply:
I've updated the PR that documents the issue and avoids problems by not applying the optimization in incremental mode.